src/main/kotlin/build/buf/intellij/index/BufModuleConfigIndex.kt

```diff
@@ -0,0 +1,109 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buf.intellij.index
+
+import build.buf.intellij.config.BufConfig
+import build.buf.intellij.module.ModuleFullName
+import com.intellij.util.indexing.DataIndexer
+import com.intellij.util.indexing.FileBasedIndex
+import com.intellij.util.indexing.FileContent
+import com.intellij.util.indexing.ID
+import com.intellij.util.indexing.ScalarIndexExtension
+import com.intellij.util.io.IOUtil
+import com.intellij.util.io.KeyDescriptor
+import org.jetbrains.yaml.YAMLUtil
+import org.jetbrains.yaml.psi.YAMLFile
+import org.jetbrains.yaml.psi.YAMLScalar
+import org.jetbrains.yaml.psi.YAMLSequence
+import java.io.DataInput
+import java.io.DataOutput
+import java.io.IOException
+
+/**
+ * [BufModuleConfigIndex] indexes [BufModuleConfig] instances loaded from `buf.yaml` files.
+ */
+class BufModuleConfigIndex : ScalarIndexExtension<BufModuleConfig>() {
+
+    override fun getName(): ID<BufModuleConfig, Void> = BufIndexes.BUF_MODULE_CONFIG_INDEX_ID
+
+    override fun dependsOnFileContent(): Boolean = true
+
+    override fun getVersion(): Int = 1
+
+    override fun getInputFilter(): FileBasedIndex.InputFilter = FileBasedIndex.InputFilter { file ->
+        file.name == BufConfig.BUF_YAML
+    }
+
+    override fun getKeyDescriptor(): KeyDescriptor<BufModuleConfig> = BufModuleConfigIndexEntryKeyDescriptor
+
+    override fun getIndexer(): DataIndexer<BufModuleConfig, Void, FileContent> {
+        return DataIndexer { inputData ->
+            val yamlFile = inputData.psiFile as? YAMLFile ?: return@DataIndexer emptyMap()
+            val bufYamlFile = yamlFile.virtualFile ?: return@DataIndexer emptyMap()
+            val modulePath = bufYamlFile.parent ?: return@DataIndexer emptyMap()
+            val version = YAMLUtil.getQualifiedKeyInFile(yamlFile, "version")?.valueText ?: return@DataIndexer emptyMap()
+            val name = YAMLUtil.getQualifiedKeyInFile(yamlFile, "name")?.valueText?.let {
+                ModuleFullName.parse(it).getOrNull()
+            }
+            val result = linkedSetOf<BufModuleConfig>()
+            when (version) {
+                "v1beta1" -> {
+                    // NOTE: No support for roots/excludes. Unsupported so far in the plugin and users should move to v1/v2.
+                    result.add(BufModuleConfig(bufYamlFile.url, modulePath.url, moduleFullName = name))
+                }
+                "v1" -> {
+                    result.add(BufModuleConfig(bufYamlFile.url, modulePath.url, moduleFullName = name))
+                }
+                "v2" -> {
+                    val modules = YAMLUtil.getQualifiedKeyInFile(yamlFile, "modules")?.value as? YAMLSequence
+                    if (modules != null && modules.items.isNotEmpty()) {
```

Comment:
This is where we have special handling to discover multiple modules in a v2 config file, and create a new `BufModuleConfig` for each one.